src/metabase/notification/seed.clj

```diff
@@ -1,124 +1,178 @@
 (ns metabase.notification.seed
   "Seed default notifications on startup.
-  Will truncate al notifications related tables then reinsert the default notifications."
+  No-op if none of the notifications are changed.
+  If a notification is changed, it will be replaced with the new one."
   (:require
-   [metabase.db :as db]
    [metabase.models.notification :as models.notification]
    [metabase.models.permissions-group :as perms-group]
+   [metabase.util :as u]
+   [metabase.util.json :as json]
    [metabase.util.log :as log]
    [toucan2.core :as t2]))
 
+(def ^:private model->compare-keys
+  "A mapping of model to the keys that are used to check if a row is the same."
+  {:model/Notification             [:payload_type :active]
+   :model/NotificationSubscription [:type :event_name :cron_schedule]
+   :model/NotificationHandler      [:channel_type :active]
+   :model/NotificationRecipient    [:type :user_id :permissions_group_id :details]
+   :model/Channel                  [:channel_type :details]
+   :model/ChannelTemplate          [:channel_type :name :details]})
+
+(defn- select-keys-nil-vals
+  "Select keys from a map, missing keys are included with value = nil."
+  [m ks]
+  (merge (into {} (zipmap ks (repeat nil)))
+         (select-keys m ks)))
+
+(defn- sanitize-model
+  [model data]
+  (select-keys-nil-vals data (get model->compare-keys model)))
+
+(defn- sanitize-notification
+  [{:keys [subscriptions handlers] :as notification}]
+  (assoc (sanitize-model :model/Notification notification)
+         :subscriptions (set (mapv #(sanitize-model :model/NotificationSubscription %) subscriptions))
+         :handlers      (set (mapv (fn [{:keys [channel template recipients] :as handler}]
+                                     (assoc (sanitize-model :model/NotificationHandler handler)
+                                            :channel    (sanitize-model :model/Channel channel)
+                                            :template   (sanitize-model :model/ChannelTemplate template)
+                                            :recipients (set (mapv #(sanitize-model :model/NotificationRecipient %) recipients))))
+                                   handlers))))
+
 (def ^:private default-notifications
+  "List of notifications that are seeded into the database on startup.
+  This list should only be modified or appended, never removed.
```

Comment:
I'm a bit worried about people missing this instruction. 

What do you think about a unit test which asserts the set of internal ids? A bit annoying to update, but more fool proof. Speaking as a fool 😄

Reply:
Oh nevermind, I see that the children elements don't have internal ids, so wouldn't really cover things.